File: service-contracts/python/contracts.py
from typing import Annotated, Dict, List, Literal, Optional, Union, get_args

from pydantic import BaseModel, ConfigDict, Field

UserRole = Literal["doctor", "patient"]
USER_ROLES = get_args(UserRole)

AppointmentStatus = Literal[
    "pending_code",
    "confirmed",
    "in_progress",
    "completed",
    "cancelled",
]
APPOINTMENT_STATUSES = get_args(AppointmentStatus)

ConsultationStatus = Literal["active", "ended"]

FeedbackSeverity = Literal["info", "warning", "critical"]

AttachmentKind = Literal["document", "image", "other"]

HealthStatus = Literal["ok", "degraded"]

DependencyStatus = Literal["up", "down", "unknown"]

ErrorCode = Literal[
    "VALIDATION_FAILED",
    "UNAUTHENTICATED",
    "FORBIDDEN",
    "NOT_FOUND",
    "CONFLICT",
    "EMAIL_TAKEN",
    "INVALID_CREDENTIALS",
    "INVALID_REFRESH_TOKEN",
    "INVALID_OR_EXPIRED_CODE",
    "TOO_MANY_ATTEMPTS",
    "RATE_LIMITED",
    "SLOT_TAKEN",
    "APPOINTMENT_NOT_CONFIRMED",
    "APPOINTMENT_TERMINAL",
    "CONSULTATION_EXISTS",
    "CONSULTATION_NOT_ACTIVE",
    "RECORD_EXISTS",
    "FILE_TOO_LARGE",
    "UNSUPPORTED_MEDIA_TYPE",
    "AI_UNAVAILABLE",
    "INTERNAL_ERROR",
]

PRE_CONSULT_QUESTION_KEYS = (
    "chief_complaint",
    "symptom_duration",
    "current_medications",
    "allergies",
    "medical_history",
)

PreConsultQuestionKey = Literal[
    "chief_complaint",
    "symptom_duration",
    "current_medications",
    "allergies",
    "medical_history",
]


class StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class ErrorResponseDto(StrictModel):
    statusCode: int
    error: str
    message: str
    details: Optional[List[str]]
    timestamp: str
    path: str
    correlationId: str


class HealthResponseDto(StrictModel):
    status: Literal["ok"]
    uptimeSeconds: float
    timestamp: str
    version: str


class ReadyChecksDto(StrictModel):
    database: DependencyStatus
    ai: DependencyStatus


class ReadyResponseDto(StrictModel):
    status: HealthStatus
    checks: ReadyChecksDto
    timestamp: str


class UserDto(StrictModel):
    id: str
    name: str
    email: str
    role: UserRole
    specialty: Optional[str]
    crm: Optional[str]
    createdAt: str


class DoctorSummaryDto(StrictModel):
    id: str
    name: str
    specialty: Optional[str]
    crm: Optional[str]


class CounterpartDto(StrictModel):
    id: str
    name: str
    role: UserRole
    specialty: Optional[str]


class UpdateUserRequestDto(StrictModel):
    name: Optional[str] = None
    specialty: Optional[str] = None


class AuthTokensDto(StrictModel):
    accessToken: str
    refreshToken: str
    tokenType: Literal["Bearer"]
    expiresIn: int


class RegisterRequestDto(StrictModel):
    name: str
    email: str
    password: str
    role: UserRole
    specialty: Optional[str] = None
    crm: Optional[str] = None


class LoginRequestDto(StrictModel):
    email: str
    password: str


class RefreshRequestDto(StrictModel):
    refreshToken: str


class LogoutRequestDto(StrictModel):
    refreshToken: Optional[str] = None


class AuthResponseDto(StrictModel):
    user: UserDto
    tokens: AuthTokensDto


class SlotDto(StrictModel):
    startsAt: str
    endsAt: str
    doctorId: str


class PreConsultAnswerInputDto(StrictModel):
    questionKey: str
    answer: str


class PreConsultAnswerDto(StrictModel):
    id: str
    appointmentId: str
    questionKey: str
    answer: str
    createdAt: str


class CreateAppointmentRequestDto(StrictModel):
    doctorId: str
    scheduledAt: str
    preConsult: Optional[List[PreConsultAnswerInputDto]] = None


class RequestCodeResponseDto(StrictModel):
    appointmentId: str
    expiresAt: str
    attemptsRemaining: int


class VerifyCodeRequestDto(StrictModel):
    code: str


class AppointmentDto(StrictModel):
    id: str
    patientId: str
    doctorId: str
    scheduledAt: str
    status: AppointmentStatus
    createdAt: str
    updatedAt: str


class CalendarEntryDto(StrictModel):
    appointmentId: str
    scheduledAt: str
    status: AppointmentStatus
    counterpart: CounterpartDto
    consultationId: Optional[str]
    isCurrent: bool


class DoctorAppointmentDto(StrictModel):
    appointmentId: str
    scheduledAt: str
    status: AppointmentStatus
    patient: CounterpartDto
    consultationId: Optional[str]
    hasPreConsult: bool


class CancelAppointmentRequestDto(StrictModel):
    reason: Optional[str] = None


class IceServerDto(StrictModel):
    urls: List[str]
    username: Optional[str] = None
    credential: Optional[str] = None


class ParticipantDto(StrictModel):
    userId: str
    role: UserRole
    joinedAt: Optional[str]


class StartConsultationResponseDto(StrictModel):
    consultationId: str
    appointmentId: str
    status: ConsultationStatus
    startedAt: str
    roomId: str
    iceServers: List[IceServerDto]
    participants: List[ParticipantDto]


class ConsultationDto(StrictModel):
    id: str
    appointmentId: str
    status: ConsultationStatus
    startedAt: str
    endedAt: Optional[str]


class EndConsultationResponseDto(StrictModel):
    consultationId: str
    appointmentId: str
    status: ConsultationStatus
    startedAt: str
    endedAt: str


class ConsultationHistoryItemDto(StrictModel):
    consultationId: str
    appointmentId: str
    startedAt: str
    endedAt: Optional[str]
    doctor: CounterpartDto
    diagnosis: Optional[str]


class CreateMedicalRecordRequestDto(StrictModel):
    consultationId: str
    patientId: str
    notes: str
    diagnosis: Optional[str] = None
    prescriptions: Optional[List[str]] = None


class MedicalRecordDto(StrictModel):
    id: str
    patientId: str
    consultationId: str
    createdBy: str
    notes: str
    diagnosis: Optional[str]
    prescriptions: List[str]
    createdAt: str


class MedicalRecordSummaryDto(StrictModel):
    id: str
    consultationId: str
    createdAt: str
    diagnosis: Optional[str]


class PatientOverviewDto(StrictModel):
    patient: UserDto
    upcomingAppointment: Optional[AppointmentDto]
    history: List[MedicalRecordSummaryDto]
    preConsult: List[PreConsultAnswerDto]


class AttachmentDto(StrictModel):
    id: str
    appointmentId: str
    consultationId: Optional[str]
    uploaderId: str
    fileName: str
    contentType: str
    sizeBytes: int
    kind: AttachmentKind
    downloadUrl: str
    createdAt: str


class CreateAiSessionRequestDto(StrictModel):
    consultationId: str
    appointmentId: str
    language: Literal["pt-BR"]


class AiSessionCreatedDto(StrictModel):
    sessionId: str
    expiresAt: str


class AiHealthDto(StrictModel):
    status: Literal["ok"]
    provider: str
    timestamp: str


class AiAudioChunkFrame(StrictModel):
    type: Literal["audio.chunk"]
    seq: int
    data: str
    encoding: Literal["pcm_s16le"]
    sampleRate: int
    channels: Literal[1]


class AiAudioEndFrame(StrictModel):
    type: Literal["audio.end"]
    seq: int


class AiSessionCloseFrame(StrictModel):
    type: Literal["session.close"]
    reason: Optional[str] = None


AiClientFrame = Annotated[
    Union[AiAudioChunkFrame, AiAudioEndFrame, AiSessionCloseFrame],
    Field(discriminator="type"),
]


class AiTranscriptPartialFrame(StrictModel):
    type: Literal["transcript.partial"]
    text: str
    at: str


class AiTranscriptFinalFrame(StrictModel):
    type: Literal["transcript.final"]
    segmentId: str
    text: str
    at: str


class AiCopilotFeedbackFrame(StrictModel):
    type: Literal["copilot.feedback"]
    severity: FeedbackSeverity
    message: str
    at: str
    tags: List[str]


class AiErrorFrame(StrictModel):
    type: Literal["error"]
    code: str
    message: str
    at: str


AiServerFrame = Annotated[
    Union[
        AiTranscriptPartialFrame,
        AiTranscriptFinalFrame,
        AiCopilotFeedbackFrame,
        AiErrorFrame,
    ],
    Field(discriminator="type"),
]

WebSocketEventPayloads: Dict[str, str] = {
    "room.joined": "appointmentId, participants, iceServers",
    "room.error": "code, message",
    "consultation.started": "appointmentId, consultationId, doctorId, startedAt",
    "consultation.ended": "appointmentId, consultationId, endedAt",
    "participant.joined": "appointmentId, userId, role, joinedAt",
    "participant.left": "appointmentId, userId, role, leftAt",
    "webrtc.offer": "appointmentId, fromUserId, sdp",
    "webrtc.answer": "appointmentId, fromUserId, sdp",
    "webrtc.ice": "appointmentId, fromUserId, candidate, sdpMid, sdpMLineIndex",
    "media.state": "appointmentId, userId, mic, camera",
    "transcript.partial": "consultationId, text, at",
    "transcript.final": "consultationId, segmentId, text, at",
    "copilot.feedback": "consultationId, severity, message, at, tags",
    "ai.status": "consultationId, status",
    "error": "code, message",
}

File: front-end/lib/contracts.ts
import { z } from "zod";
import {
  APPOINTMENT_STATUSES,
  PRE_CONSULT_QUESTION_KEYS,
  USER_ROLES,
  type AttachmentKind as ContractAttachmentKind,
  type ConsultationStatus as ContractConsultationStatus,
  type DependencyStatus as ContractDependencyStatus,
  type ErrorCode as ContractErrorCode,
  type FeedbackSeverity as ContractFeedbackSeverity,
  type HealthStatus as ContractHealthStatus,
} from "@telemed/service-contracts";

export { APPOINTMENT_STATUSES, PRE_CONSULT_QUESTION_KEYS, USER_ROLES };
export type UserRole = (typeof USER_ROLES)[number];
export const userRoleSchema = z.enum(USER_ROLES);

export type AppointmentStatus = (typeof APPOINTMENT_STATUSES)[number];
export const appointmentStatusSchema = z.enum(APPOINTMENT_STATUSES);

export const CONSULTATION_STATUSES = ["active", "ended"] as const satisfies readonly ContractConsultationStatus[];
export type ConsultationStatus = (typeof CONSULTATION_STATUSES)[number];
export const consultationStatusSchema = z.enum(CONSULTATION_STATUSES);

export const FEEDBACK_SEVERITIES = ["info", "warning", "critical"] as const satisfies readonly ContractFeedbackSeverity[];
export type FeedbackSeverity = (typeof FEEDBACK_SEVERITIES)[number];
export const feedbackSeveritySchema = z.enum(FEEDBACK_SEVERITIES);

export const ATTACHMENT_KINDS = ["document", "image", "other"] as const satisfies readonly ContractAttachmentKind[];
export type AttachmentKind = (typeof ATTACHMENT_KINDS)[number];
export const attachmentKindSchema = z.enum(ATTACHMENT_KINDS);

export const HEALTH_STATUSES = ["ok", "degraded"] as const satisfies readonly ContractHealthStatus[];
export const healthStatusSchema = z.enum(HEALTH_STATUSES);

export const DEPENDENCY_STATUSES = ["up", "down", "unknown"] as const satisfies readonly ContractDependencyStatus[];
export const dependencyStatusSchema = z.enum(DEPENDENCY_STATUSES);

export const ERROR_CODES = [
  "VALIDATION_FAILED",
  "UNAUTHENTICATED",
  "FORBIDDEN",
  "NOT_FOUND",
  "CONFLICT",
  "EMAIL_TAKEN",
  "INVALID_CREDENTIALS",
  "INVALID_REFRESH_TOKEN",
  "INVALID_OR_EXPIRED_CODE",
  "TOO_MANY_ATTEMPTS",
  "RATE_LIMITED",
  "SLOT_TAKEN",
  "APPOINTMENT_NOT_CONFIRMED",
  "APPOINTMENT_TERMINAL",
  "CONSULTATION_EXISTS",
  "CONSULTATION_NOT_ACTIVE",
  "RECORD_EXISTS",
  "FILE_TOO_LARGE",
  "UNSUPPORTED_MEDIA_TYPE",
  "AI_UNAVAILABLE",
  "INTERNAL_ERROR",
] as const satisfies readonly ContractErrorCode[];
export type ErrorCode = (typeof ERROR_CODES)[number];
export const errorCodeSchema = z.enum(ERROR_CODES);

export const errorResponseSchema = z.object({
  statusCode: z.number().int(),
  error: z.string(),
  message: z.string(),
  details: z.array(z.string()).nullable(),
  timestamp: z.string(),
  path: z.string(),
  correlationId: z.string(),
});
export type ErrorResponseDto = z.infer<typeof errorResponseSchema>;

export const healthResponseSchema = z.object({
  status: z.literal("ok"),
  uptimeSeconds: z.number(),
  timestamp: z.string(),
  version: z.string(),
});
export type HealthResponseDto = z.infer<typeof healthResponseSchema>;

export const readyResponseSchema = z.object({
  status: healthStatusSchema,
  checks: z.object({
    database: dependencyStatusSchema,
    ai: dependencyStatusSchema,
  }),
  timestamp: z.string(),
});
export type ReadyResponseDto = z.infer<typeof readyResponseSchema>;

export const userDtoSchema = z.object({
  id: z.string(),
  name: z.string(),
  email: z.email(),
  role: userRoleSchema,
  specialty: z.string().nullable(),
  crm: z.string().nullable(),
  createdAt: z.string(),
});
export type UserDto = z.infer<typeof userDtoSchema>;

export const doctorSummaryDtoSchema = z.object({
  id: z.string(),
  name: z.string(),
  specialty: z.string().nullable(),
  crm: z.string().nullable(),
});
export type DoctorSummaryDto = z.infer<typeof doctorSummaryDtoSchema>;

export const counterpartDtoSchema = z.object({
  id: z.string(),
  name: z.string(),
  role: userRoleSchema,
  specialty: z.string().nullable(),
});
export type CounterpartDto = z.infer<typeof counterpartDtoSchema>;

export const updateUserRequestSchema = z.object({
  name: z.string().min(2).max(120).optional(),
  specialty: z.string().min(2).max(120).optional(),
});
export type UpdateUserRequestDto = z.infer<typeof updateUserRequestSchema>;

export const authTokensSchema = z.object({
  accessToken: z.string(),
  refreshToken: z.string(),
  tokenType: z.literal("Bearer"),
  expiresIn: z.number().int(),
});
export type AuthTokensDto = z.infer<typeof authTokensSchema>;

export const registerRequestSchema = z.object({
  name: z.string().min(2, "Informe o nome completo").max(120),
  email: z.email("Informe um e-mail válido"),
  password: z.string().min(8, "Use ao menos 8 caracteres").max(72),
  role: userRoleSchema,
  specialty: z.string().min(2).max(120).optional(),
  crm: z.string().min(2).max(40).optional(),
});
export type RegisterRequestDto = z.infer<typeof registerRequestSchema>;

export const loginRequestSchema = z.object({
  email: z.email("Informe um e-mail válido"),
  password: z.string().min(1, "Informe a senha").max(72),
});
export type LoginRequestDto = z.infer<typeof loginRequestSchema>;

export const refreshRequestSchema = z.object({
  refreshToken: z.string().min(1),
});
export type RefreshRequestDto = z.infer<typeof refreshRequestSchema>;

export const logoutRequestSchema = z.object({
  refreshToken: z.string().min(1).optional(),
});
export type LogoutRequestDto = z.infer<typeof logoutRequestSchema>;

export const authResponseSchema = z.object({
  user: userDtoSchema,
  tokens: authTokensSchema,
});
export type AuthResponseDto = z.infer<typeof authResponseSchema>;

export const preConsultQuestionKeySchema = z.enum(PRE_CONSULT_QUESTION_KEYS);
export type PreConsultQuestionKey = z.infer<typeof preConsultQuestionKeySchema>;

export const preConsultAnswerInputSchema = z.object({
  questionKey: preConsultQuestionKeySchema,
  answer: z.string(),
});
export type PreConsultAnswerInputDto = z.infer<typeof preConsultAnswerInputSchema>;

export const preConsultAnswerSchema = z.object({
  id: z.string(),
  appointmentId: z.string(),
  questionKey: z.string(),
  answer: z.string(),
  createdAt: z.string(),
});
export type PreConsultAnswerDto = z.infer<typeof preConsultAnswerSchema>;

export const slotSchema = z.object({
  startsAt: z.string(),
  endsAt: z.string(),
  doctorId: z.string(),
});
export type SlotDto = z.infer<typeof slotSchema>;

export const createAppointmentRequestSchema = z.object({
  doctorId: z.string(),
  scheduledAt: z.string(),
  preConsult: z.array(preConsultAnswerInputSchema).optional(),
});
export type CreateAppointmentRequestDto = z.infer<typeof createAppointmentRequestSchema>;

export const requestCodeResponseSchema = z.object({
  appointmentId: z.string(),
  expiresAt: z.string(),
  attemptsRemaining: z.number().int(),
});
export type RequestCodeResponseDto = z.infer<typeof requestCodeResponseSchema>;

export const verifyCodeRequestSchema = z.object({
  code: z.string(),
});
export type VerifyCodeRequestDto = z.infer<typeof verifyCodeRequestSchema>;

export const appointmentSchema = z.object({
  id: z.string(),
  patientId: z.string(),
  doctorId: z.string(),
  scheduledAt: z.string(),
  status: appointmentStatusSchema,
  createdAt: z.string(),
  updatedAt: z.string(),
});
export type AppointmentDto = z.infer<typeof appointmentSchema>;

export const calendarEntrySchema = z.object({
  appointmentId: z.string(),
  scheduledAt: z.string(),
  status: appointmentStatusSchema,
  counterpart: counterpartDtoSchema,
  consultationId: z.string().nullable(),
  isCurrent: z.boolean(),
});
export type CalendarEntryDto = z.infer<typeof calendarEntrySchema>;

export const doctorAppointmentSchema = z.object({
  appointmentId: z.string(),
  scheduledAt: z.string(),
  status: appointmentStatusSchema,
  patient: counterpartDtoSchema,
  consultationId: z.string().nullable(),
  hasPreConsult: z.boolean(),
});
export type DoctorAppointmentDto = z.infer<typeof doctorAppointmentSchema>;

export const cancelAppointmentRequestSchema = z.object({
  reason: z.string().optional(),
});
export type CancelAppointmentRequestDto = z.infer<typeof cancelAppointmentRequestSchema>;

export const iceServerSchema = z.object({
  urls: z.array(z.string()),
  username: z.string().optional(),
  credential: z.string().optional(),
});
export type IceServerDto = z.infer<typeof iceServerSchema>;

export const participantSchema = z.object({
  userId: z.string(),
  role: userRoleSchema,
  joinedAt: z.string().nullable(),
});
export type ParticipantDto = z.infer<typeof participantSchema>;

export const startConsultationResponseSchema = z.object({
  consultationId: z.string(),
  appointmentId: z.string(),
  status: consultationStatusSchema,
  startedAt: z.string(),
  roomId: z.string(),
  iceServers: z.array(iceServerSchema),
  participants: z.array(participantSchema),
});
export type StartConsultationResponseDto = z.infer<typeof startConsultationResponseSchema>;

export const consultationSchema = z.object({
  id: z.string(),
  appointmentId: z.string(),
  status: consultationStatusSchema,
  startedAt: z.string(),
  endedAt: z.string().nullable(),
});
export type ConsultationDto = z.infer<typeof consultationSchema>;

export const endConsultationResponseSchema = z.object({
  consultationId: z.string(),
  appointmentId: z.string(),
  status: consultationStatusSchema,
  startedAt: z.string(),
  endedAt: z.string(),
});
export type EndConsultationResponseDto = z.infer<typeof endConsultationResponseSchema>;

export const consultationHistoryItemSchema = z.object({
  consultationId: z.string(),
  appointmentId: z.string(),
  startedAt: z.string(),
  endedAt: z.string().nullable(),
  doctor: counterpartDtoSchema,
  diagnosis: z.string().nullable(),
});
export type ConsultationHistoryItemDto = z.infer<typeof consultationHistoryItemSchema>;

export const createMedicalRecordRequestSchema = z.object({
  consultationId: z.string(),
  patientId: z.string(),
  notes: z.string(),
  diagnosis: z.string().optional(),
  prescriptions: z.array(z.string()).optional(),
});
export type CreateMedicalRecordRequestDto = z.infer<typeof createMedicalRecordRequestSchema>;

export const medicalRecordSchema = z.object({
  id: z.string(),
  patientId: z.string(),
  consultationId: z.string(),
  createdBy: z.string(),
  notes: z.string(),
  diagnosis: z.string().nullable(),
  prescriptions: z.array(z.string()),
  createdAt: z.string(),
});
export type MedicalRecordDto = z.infer<typeof medicalRecordSchema>;

export const medicalRecordSummarySchema = z.object({
  id: z.string(),
  consultationId: z.string(),
  createdAt: z.string(),
  diagnosis: z.string().nullable(),
});
export type MedicalRecordSummaryDto = z.infer<typeof medicalRecordSummarySchema>;

export const patientOverviewSchema = z.object({
  patient: userDtoSchema,
  upcomingAppointment: appointmentSchema.nullable(),
  history: z.array(medicalRecordSummarySchema),
  preConsult: z.array(preConsultAnswerSchema),
});
export type PatientOverviewDto = z.infer<typeof patientOverviewSchema>;

export const attachmentSchema = z.object({
  id: z.string(),
  appointmentId: z.string(),
  consultationId: z.string().nullable(),
  uploaderId: z.string(),
  fileName: z.string(),
  contentType: z.string(),
  sizeBytes: z.number().int(),
  kind: attachmentKindSchema,
  downloadUrl: z.string(),
  createdAt: z.string(),
});
export type AttachmentDto = z.infer<typeof attachmentSchema>;

export const doctorAppointmentListSchema = z.array(doctorAppointmentSchema);
export const preConsultAnswerListSchema = z.array(preConsultAnswerSchema);
export const consultationHistoryListSchema = z.array(consultationHistoryItemSchema);
export const medicalRecordSummaryListSchema = z.array(medicalRecordSummarySchema);
